Add doc for Recovery from expansion failure

gnufied opened the pull request:
xref https://github.com/kubernetes/enhancements/issues/1790

mehabhalodiya commented:
/assign @ramrodo

tengqm commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -165,6 +165,7 @@ different Kubernetes components.
 | `ProxyTerminatingEndpoints` | `false` | Alpha | 1.22 | |
 | `QOSReserved` | `false` | Alpha | 1.11 | |
 | `ReadWriteOncePod` | `false` | Alpha | 1.22 | |
+| `RecoverVolumeExpansionFailure` | `false` | Alpha | 1.23 | |
```
In addition to this, we expect a brief introduction (possibly with a link to details) added at the end of this page.

gnufied replied:
added more details.

sftim replied:
I don't see the extra details. This should be in the form of an edit to the page section headed “List of feature gates”.

gnufied replied:
Added details below. PTAL.

sftim reviewed (commented):
Hi @gnufied 

You should also document the feature. Unless that's in another PR, I want to highlight that this new feature isn't documented.

/hold

gnufied commented:
Sorry, this was placeholder PR. I will update the docs.

xing-yang commented:
/assign

sftim commented:
/sig storage

sftim commented on content/en/docs/reference/access-authn-authz/admission-controllers.md:
```diff
@@ -583,7 +583,7 @@ subresource of the referenced *owner* can change it.
 
 {{< note >}}
-Support for volume resizing is available as an alpha feature. Admins must set the feature gate `ExpandPersistentVolumes`
+Support for volume resizing is available as an beta feature. Admins must set the feature gate `ExpandPersistentVolumes`
```
```suggestion
Support for volume resizing is available as a beta feature. As a cluster administrator,
you must ensure that the feature gate `ExpandPersistentVolumes` is set
```

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -307,6 +307,18 @@ If expanding underlying storage fails, the cluster administrator can manually re
 4. Re-create the PVC with smaller size than PV and set `volumeName` field of the PVC to the name of the PV. This should bind new PVC to existing PV.
 5. Don't forget to restore the reclaim policy of the PV.
 
+##### User recovery expansion failure
```
It looks like there are two ways to attempt recovery when expansion fails.

I would use https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#tabs to show the two choices.

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -307,6 +307,18 @@ If expanding underlying storage fails, the cluster administrator can manually re
+Recovery from expanding PVCs by users is available as an alpha feature since Kubernetes 1.23. The `RecoverVolumeExpansionFailure` feature must be enabled for this feature to work. Refer to the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) documentation for more information.
+{{< /note >}}
+
+In addition to above manual steps Kubernetes if feature gate `RecoverVolumeExpansionFailure` is enabled in the cluster, the user of PVC can retry expansion with smaller size than previously requested value by editing `pvc.spec.resources`.
```
```suggestion
If the feature gates `ExpandPersistentVolumes` and `RecoverVolumeExpansionFailure` are both
enabled in your cluster, and expansion has failed for a PVC, you can retry expansion with a
smaller size than the previously requested value. To request a new expansion attempt with a
smaller proposed size, edit `.spec.resources` for that PVC and choose a value that is less than the
value you previously tried.
```

ramrodo replied:
```suggestion
In addition to the above manual steps if feature gate `RecoverVolumeExpansionFailure` is enabled in the cluster, the user of PVC can retry expansion with a smaller size than previously requested value by editing `pvc.spec.resources`.
```

I think the word `Kubernetes` is left over. If not, please ignore this comment.

xing-yang replied:
manual steps Kubernetes -> manual steps in Kubernetes

ramrodo commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -307,6 +307,18 @@ If expanding underlying storage fails, the cluster administrator can manually re
+{{< /note >}}
+
+In addition to above manual steps Kubernetes if feature gate `RecoverVolumeExpansionFailure` is enabled in the cluster, the user of PVC can retry expansion with smaller size than previously requested value by editing `pvc.spec.resources`.
+This is useful if expansion to a higher value can not succeed because of capacity constraints, in which case user of PVC can retry expansion by specifying a size that is within the capacity limits of underlying storage provider. It should be noted that, while users can specify lower value than what was requested previously in `pvc.spec.resources` but new value must still be higher than `pvc.statu ...
```
```suggestion
This is useful if expansion to a higher value can not succeed because of capacity constraints, in which case user of PVC can retry expansion by specifying a size that is within the capacity limits of the underlying storage provider. It should be noted that, while users can specify a lower value than what was requested previously in `pvc.spec.resources` but new value must still be higher than `pvc.status.capacity`.
```

gnufied replied:
Looks like you and @sftim provided conflicting suggested change, if the new change looks good let me know.

ramrodo reviewed (commented):
Just small suggestions

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -307,6 +307,18 @@ If expanding underlying storage fails, the cluster administrator can manually re
+{{< /note >}}
+
+In addition to above manual steps Kubernetes if feature gate `RecoverVolumeExpansionFailure` is enabled in the cluster, the user of PVC can retry expansion with smaller size than previously requested value by editing `pvc.spec.resources`.
+This is useful if expansion to a higher value can not succeed because of capacity constraints, in which case user of PVC can retry expansion by specifying a size that is within the capacity limits of underlying storage provider. It should be noted that, while users can specify lower value than what was requested previously in `pvc.spec.resources` but new value must still be higher than `pvc.statu ...
```
```suggestion
This is useful if expansion to a higher value did not succeed because of capacity constraint.
If that has happened, or you suspect that it might have, you can retry expansion by specifying a
size that is within the capacity limits of underlying storage provider.

Note that,
although you can a specify a lower amount of storage than what was requested previously,
the new value must still be higher than `.status.capacity`.
Kubernetes does not support shrinking a PVC to less than its current size.
```

sftim reviewed (changes requested):
At the moment this documentation isn't clear and I think it would confuse readers.

It looks like there are two ways to attempt recovery when expansion fails.

I would use https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#tabs to show the two choices.

How would someone know to change `.spec.resources` for a PVC that failed to expend? What would they do to examine the PVC and spot the failure?
(eg: `kubectl get event`?)

After you've updated @gnufied, SIG Docs can re-review.

xing-yang commented:
@gnufied Can you squash the commits when addressing the comments?

gnufied commented:
@sftim @ramrodo I incorporated most of your feedback and also added tabs for both the methods. 

> How would someone know to change .spec.resources for a PVC that failed to expend? What would they do to examine the PVC and spot the failure?

If a user knows how to expand the PVC then they know how to edit `spec.resources` ? Also added a note about how to monitor progress of resize operation.

xing-yang commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -299,6 +299,11 @@ Expanding EBS volumes is a time-consuming operation. Also, there is a per-volume
 
 #### Recovering from Failure when Expanding Volumes
 
+If a user specifies a new size that is too big to be satisfied by underlying storage system, expansion of PVC will be continuously retried until user or cluster administrator takes some action. This can be undesirable and hence  Kubernetes provides following methods of recovering from such failures.
```
nit: extra white spaces between "hence" and "Kubernetes".

gnufied replied:
fixed

sftim commented:
/hold cancel

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -889,6 +890,8 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `ReadWriteOncePod`: Enables the usage of `ReadWriteOncePod` PersistentVolume
   access mode.
+- `RecoverVolumeExpansionFailure`: Enables users to edit their PVCs to smaller sizes so as they can recover from previously issued
+  volume expansion failures. See [enhancement proposal](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1790-recover-resize-failure/README.md) for more details.
```
```suggestion
  volume expansion failures. See
  [Recovering from Failure when Expanding Volumes](/docs/concepts/storage/persistent-volumes/#recovering-from-failure-when-expanding-volumes)
  for more details.
```
?
(We prefer to link to docs rather than KEPs)

xing-yang commented:
/lgtm

ramrodo commented:
/lgtm
cc @jlbutler

jlbutler commented:
/approve
